curl/tests/mock_curl.hpp

```diff
@@ -0,0 +1,42 @@
+#pragma once
+#ifdef _WIN32
+#    pragma GCC diagnostic ignored "-Wattributes"
+#    define CURL_STATICLIB
+#endif
+
+#include <string>
+#include <functional>
+#include <curl/curl.h>
+
+struct curl_impl
+{
+    enum struct http_method
+    {
+        get,
+        put,
+        post
+    };
+
+    // Pointer for client::write_header as a callback function in curl_easy_setopt
+    std::function<size_t(char*, size_t, size_t, void*)> write_header;
+    void* header_context; // Where to write the header part of the received data to
+
+    // Pointer for client::write_body as a callback function in curl_easy_setopt
+    std::function<size_t(char*, size_t, size_t, void*)> write_body;
+    void* body_context; // Where to write the body part of the received data to
+
+    // Pointer for client::read_body as a callback function in curl_easy_setopt
+    std::function<size_t(char*, size_t, size_t, void*)> read_function;
+    void* read_data; // Where to read the request body from
+
+    std::string request_url, cookie, cacert, client_cert, client_key;
+    http_method method = http_method::get;
+
+    curl_slist* header; // List of custom request headers to be passed to the server
+
+    std::string read_buffer; // Buffer to test reading the request body
+    std::string resp_body;   // Response body which should be written to a context using the write_body function callback
+};
+
+void set_error_mode(bool mode);
+static bool error_mode = false;
```

Comment:
The error_mode flag should only exist in `mock_curl.cc`.

The only need we have for a global is when the curl_impl struct doesn't exist, i.e. during creation of the client. We use this to signal that an error should occur during client initialization.

I also lean towards making this a struct, so we can take advantage of RAII for automatic cleanup after the test.